modules/sfp_countryname.py

```diff
@@ -0,0 +1,323 @@
+        tldNonCountryCodes = self.getNonCountryCodesDict()
+
+        # Split domain into parts by '.'
+        # We can conclude that country TLDs are reserved 
+        domainParts = srcDomain.split(".")
+
+        # Search for country TLD in the domain parts - reversed
+        for part in domainParts[::-1]:
+            if part.upper() in tldCountryCodes.keys():
+                return tldCountryCodes[part.upper()]
+
+        # Search for non country TLD in the domain parts
+        for part in domainParts[::-1]:
+            if part.upper() in tldNonCountryCodes.keys():
+                return tldNonCountryCodes[part.upper()]
+
+        # No associated country name is found
+        return None
+
+    # Detect name of country from IBAN
+    def detectCountryFromIBAN(self, srcIBAN):
+
+        # Get dictionary of country codes and country names
+        tldCountryCodes = self.getCountryCodeDict()
+        try:
+            # Get country code from IBAN 
+            countryCode = srcIBAN[0:2].upper()
+            return tldCountryCodes[countryCode]
+        except:
+            # No country name is found in the IBAN
+            return None
+    
+    # Detect name of country from Who Is lookup data
+    def detectCountryFromWhoIs(self, srcWhoIs):
+        
+        # Get dictionary of country codes and  country names
+        abbvCountryCodes = self.getCountryCodeDict()
+        countries = set()
+
+        # Look for countrycodes and country in whois data
+        for countryName in abbvCountryCodes.values(): 
+
+            # Look for country name in whois data
+            matchCountries = re.findall("[\s,'\"]" + countryName + "[\s,'\"]", srcWhoIs, re.IGNORECASE)
+
+            if len(matchCountries) > 0:
+                # Get country name from first index of list
+                # Extract only the text part of the country code
+                matchCountry = matchCountries[0].strip(",").strip("'").strip("\"").strip()
+                countries.add(matchCountry)
+        
+        # If any countries are found
+        if len(countries) > 0:
+            return countries
+
+        return None
+            
+    # What events is this module interested in for input
+    def watchedEvents(self):
+        return ["IBAN_NUMBER", "PHONE_NUMBER", "AFFILIATE_DOMAIN_NAME",
+                "CO_HOSTED_SITE_DOMAIN", "DOMAIN_NAME", "SIMILARDOMAIN",
+                "AFFILIATE_DOMAIN_WHOIS", "CO_HOSTED_SITE_DOMAIN_WHOIS",
+                'DOMAIN_WHOIS']
+
+    # What events this module produces
+    def producedEvents(self):
+        return ["COUNTRY_NAME"]
+
+    # Handle events sent to this module
+    def handleEvent(self, event):
+        eventName = event.eventType
+        srcModuleName = event.module
+        eventData = event.data
+        
+        evttype = "COUNTRY_NAME"
+
+        self.sf.debug("Received event, " + eventName + ", from " + srcModuleName)
+
+        myres = list()
+        countryName = None
+
```

Comment:
Duplication check needed here, to not process events for anything already processed.